.github/workflows/Tests.yaml

```diff
@@ -433,3 +446,86 @@ jobs:
           # We get occasional random timeouts, repeat tests to see if
           # it is a random timeout or systematic
           ctest -j2 --output-on-failure -L inputfiles --repeat after-timeout:3
+
+  # Release a new version on scheduled or manual events when the tests pass.
+  # Only enable this on the `sxs-collaboration/spectre` repository (not on
+  # forks).
+  release_version:
+    name: Release version
+    if: >
+      github.repository == 'sxs-collaboration/spectre' &&
+        (github.event_name == 'schedule'
+          || (github.event_name == 'workflow_dispatch'
+            && github.event.inputs.release_version != ''))
+    needs:
+      - check_files_and_formatting
+      - doc_check
+      - unit_tests
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - uses: actions/setup-python@v2
+        with:
+          python-version: '3.x'
+      - name: Install Python dependencies
+        run: |
+          pip install GitPython PyGithub tqdm
+      # We use the current date as tag name, unless a tag name was specified
+      # as input to the `workflow_dispatch` event
+      - name: Determine release version
+        id: get_version
+        run: |
+          INPUT_RELEASE_VERSION=${{ github.event.inputs.release_version }}
+          RELEASE_VERSION=${INPUT_RELEASE_VERSION:-$(date +'%Y-%m-%d')}
+          echo "Release version is: ${RELEASE_VERSION}"
+          echo "RELEASE_VERSION=$RELEASE_VERSION" >> $GITHUB_ENV
+      - name: Validate release version
+        run: |
+          VERSION_PATTERN="^[-.0-9a-zA-Z]+$"
```

Comment:
- Yes to narrowing down the pattern
- Are you suggesting to require that the year and month in the manually-specified tag match today's year and month? Can't we trust ourselves to use the power to choose the version name responsibly? I want to avoid situations where we want to fix something but have to jump through hoops because we were too restrictive here. For example, if today is Jan 1st and we forgot to release some version before the end of the year, think we should just be able to backdate it (the branch `HEAD` is the same, after all). I could even add the option to specify the revision to tag, instead of the default `HEAD`.
- Letters are typically used for pre-release versions like `1.2.3-alpha`. Again, I think we should trust ourselves and not be too restrictive here. It's hard to anticipate what we want to do in the future.
- I used dashes instead of dots because I was worried that `YYYY.MM.DD` could be mis-interpreted as semantic versioning, implying some sort of compatibility guarantee. In fact cmake reads the components as "major", "minor" and "patch". Do you think that's not an issue?

Reply:
- :+1:
- Yep, that's what I'm suggesting. I'm not sure we should be backdating tags like that. This will break horribly if we had any PR merged between the backdate and the moment the tag is applied, I think. Or at least I'm not at all convinced this approach is safe. That might just be an issue with allowing a nonlinear history, though. @wthrowe do you have any experience with this? Should we switch to a linear history to make `checkout tag` not do something weird where we get "newer" commits in the tag? I have no experience with this.
- Since we are not using semantic versioning your example doesn't apply. No, I don't think we should trust ourselves. Everything we can check we should check.
- We should use dots, that's super standard. I don't think it's really an issue what CMake calls them, and we can't really change what CMake calls things. In a way it's actually not "too incorrect" what CMake calls them, but luckily we won't need to deal with that ever really anyway aside from filling things into some of the `cpp` files and the documentation, where all we really care about is what values they hold.

Reply:
Ok to dots. How about we don't zero-pad the month and day then, and even leave out the day on the standard scheduled monthly releases: `v2020.11`? I don't care much if we allow letters in the "tweak" component, I lean towards allowing things like `v2020.11.1.1-alpha` but am fine either way.

On backdating: Say we have `develop` point to commit `A`, which is a merge commit from a PR merge on Dec 31st, 2020. We planned on releasing that commit as `v2020.12.31` for whatever reason. But we forgot to do that on Dec 31st, because, you know, we might not have spent that particular day coding :) So I think on Jan 1st (or later) we should be able to just go on GitHub, click the "Dispatch Workflow" button, put in `v2020.12.31` as the tag name and `A` as the revision and hit "Run" to create that release.

Reply:
> do you have any experience with this? Should we switch to a linear history to make checkout tag not do something weird where we get "newer" commits in the tag? I have no experience with this.

I have no personal experience, but I think it will be OK without us changing anything.  Going backwards in history from the tag should look normal.  We may get slightly odd-looking things like a tag containing commits that are older than the previous tag, but not contained in it, but that shouldn't actually matter.  If we want to backdate, I think it is still fine as long as the old tag is reachable from the new one, which should be true for any PR merge commit.

> We should use dots

Agreed.  No one will confuse us as being on semver 2000+.

> How about we don't zero-pad the month and day then

Ew.  (And semver doesn't zero-pad, so this would seem to make confusion more likely.)

>  even leave out the day on the standard scheduled monthly releases

This seems reasonable.  We should make sure it's before any manual tags during the month, but I don't see an issue with that.

> On backdating: Say we have develop point to commit A, which is a merge commit from a PR merge on Dec 31st, 2020. We planned on releasing that commit as v2020.12.31 for whatever reason. But we forgot to do that on Dec 31st, because, you know, we might not have spent that particular day coding :) So I think on Jan 1st (or later) we should be able to just go on GitHub, click the "Dispatch Workflow" button, put in v2020.12.31 as the tag name and A as the revision and hit "Run" to create that release.

But why would we want to do that?  Is someone going to rely on the name of a tag we haven't created yet?  There's no reason to write the script to make it easier to do something if there's no reason to do that thing.

Reply:
Re version format: Do we agree on `YYYY.0M[.0D[.TWEAK]]`? I.e., a standard scheduled release would be `2021.01`, an extra release that month would be `2021.01.15` and another (unlikely) release on the same day would be `2021.01.15.1`.

Reply:
I agree on the formatting padding with zeros.

I would prefer the standard release including the date so it's clear which day that "standard" is released. Otherwise I worry that it'll be unclear if `2020.10.15` is older or newer than `2020.10` since we could be tagging at the end of the month. While this should all be documented, I don't see the drawback to being explicit here, but maybe I'm missing something.

Reply:
- By standard versioning rules `2020.10.15` would be newer than `2020.10`
- I expect we will very rarely have extra releases within a month. If feel like the more concise `2020.10` promotes the idea of this being "the October release", which is easy to refer to when talking to people.

Reply:
@nilsdeppe I'll write up the rules that we end up with, and a description how the release schedule is automated, in a dev guide

Reply:
why can't we just refer to 2020.10.01 as the Oct. release? It's just as easy to say and retains the additional info? I don't think there are "standard versioning rules" really, especially not for by date. Often the extra `.0.0` is included with semantic versioning, so why should we cut out the extra info?

Reply:
ok sure fine with me